warpgate-protocol-http/src/middleware/cookie_host.rs

```diff
@@ -38,25 +38,119 @@ impl<E: Endpoint> Endpoint for CookieHostMiddlewareEndpoint<E> {
+            .map(|h| h.split(':').next().unwrap_or(h).to_string())
+            .or_else(|| req.original_uri().host().map(|x| x.to_string()));
+
+        let scheme_https = req.original_uri().scheme() == Some(&Scheme::HTTPS);
+        let header_https = req.header("x-forwarded-proto").map(|h| h == "https").unwrap_or(false);
+        let is_https = scheme_https || header_https;
 
         let mut resp = self.inner.call(req).await?.into_response();
 
         if let Some(host) = host {
-            if let Entry::Occupied(mut entry) = resp.headers_mut().entry(http::header::SET_COOKIE) {
-                if let Ok(cookie_str) = entry.get().to_str() {
-                    if let Ok(mut cookie) = Cookie::parse(cookie_str) {
-                        if cookie.name() == SESSION_COOKIE_NAME {
-                            if let Some(ref base) = self.base_domain {
-                                // Preferred: scope cookie to parent domain for all subdomains
-                                cookie.set_domain(base);
-                            } else {
-                                // Fallback: preserve previous behavior (use request host)
-                                cookie.set_domain(&host);
+            // Extract all Set-Cookie headers for modification
+            let cookie_values: Vec<String> = {
+                let headers = resp.headers();
+                headers
+                    .get_all(http::header::SET_COOKIE)
+                    .iter()
+                    .filter_map(|v| v.to_str().ok())
+                    .map(|s| s.to_string())
+                    .collect()
+            };
+
+            // Use the cookie crate to parse and modify cookies properly
+            let mut modified_session_cookie: Option<String> = None;
+            for cookie_str in &cookie_values {
+                if let Ok(mut cookie) = Cookie::parse(cookie_str) {
+                    if cookie.name() == SESSION_COOKIE_NAME {
+                        // For localhost/127.0.0.1, omit Domain attribute since browsers won't send cookies with a different domain
+                        let is_localhost = is_localhost_host(&host);
+                        let target_domain = if is_localhost {
+                            None // Omit Domain attribute for localhost - browser will scope to exact host
+                        } else if let Some(ref base) = self.base_domain {
+                            Some(base.clone())
+                        } else {
+                            Some(host.clone())
+                        };
+
+                        // Set or remove Domain attribute using cookie crate methods
+                        if let Some(ref domain) = target_domain {
+                            cookie.set_domain(domain.clone());
+                        } else {
+                            // For localhost, we need to remove the domain attribute
+                            // Rebuild the cookie without domain, copying all other attributes
```

Comment:
If this is just about removing the Domain attribute, you can use unset_domain: https://docs.rs/cookie/latest/cookie/struct.Cookie.html#method.unset_domain